### PR TITLE
[Feature] Support nn modules for bipartite graphs.

### PR DESCRIPTION
**Pull request by yzh119**
## Description
Similar to #1297 and #1385 , this PR aims to support more nn modules for bipartite graphs.

### Progress Tracker

Models
- [x] agnnconv [@yzh119]
- [ ] appnpconv [@yzh119] (not inclined, multiple message-massing)
- [ ] atomicconv [@yzh119] (not inclined)
- [ ] cfconv [@yzh119] (not inclined)
- [ ] chebconv (not inclined, multiple message-passing)
- [ ] densechebconv (not inclined)
- [x] densegraphconv [@yzh119]
- [x] densesageconv [@yzh119]
- [x] edgeconv [@BarclayII]
- [x] gatconv [@yzh119]
- [x] ~~gatedgraphconv~~ (not inclined for multiple message passing, should have a separate GGNN-hetero module). [@BarclayII]
- [x] ginconv [@yzh119]
- [x] gmmconv [@BarclayII]
- [x] graphconv [@jermainewang ]
- [x] nnconv [@BarclayII]
- [x] ~~relgraphconv~~ (not inclined: should have a separate RGCN-hetero module) [@BarclayII]
- [x] sageconv [@BarclayII ]
- [x] ~~tagconv~~ (not inclined for multiple message passing) [@BarclayII]

Others
- [x] Unit test
- [x] Documentation
- [x] MXnet/Tensorflow

### Items to discuss
- @jermainewang , I noticed that we distinguish `D_{src}` and `D_{dst}` in the docstring of new SAGEConv interface, I wonder should we distinguish bipartite/homo graph in `__init__` function so that the network parameters would differ if `D_{src}` does not equal `D_{dst}`.
- @BarclayII , if you have time feel free to complete the missing items mentioned above.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

File: tests/mxnet/test_nn.py [hunk 1]
```diff
@@ -248,52 +263,70 @@ def test_dense_cheb_conv():
...
-    dense_conv.weight.set_data(
-        conv.weight.data())
-    dense_conv.bias.set_data(
-        conv.bias.data())
-    feat = F.randn((100, 5))
-
-    out_conv = conv(g, feat)
-    out_dense_conv = dense_conv(adj, feat)
-    assert F.allclose(out_conv, out_dense_conv)
+    for norm_type in ['both', 'right', 'none']:
```
Comment (jermainewang) [inline tests/mxnet/test_nn.py]: Use `pytest.mark.parametrize` could save you some efforts. https://github.com/dmlc/dgl/blob/master/tests/pytorch/test_nn.py#L72

File: tests/mxnet/test_nn.py [hunk 2]
```diff
@@ -248,52 +263,70 @@ def test_dense_cheb_conv():
...
-    dense_sage.fc.bias.set_data(
-        sage.fc_neigh.bias.data())
-    feat = F.randn((100, 5))
-
-    out_sage = sage(g, feat)
-    out_dense_sage = dense_sage(adj, feat)
-    assert F.allclose(out_sage, out_dense_sage)
+    for g in [
+        dgl.graph(sp.sparse.random(100, 100, density=0.1)),
+        dgl.bipartite(sp.sparse.random(100, 200, density=0.1))]:
```
Comment (jermainewang) [inline tests/mxnet/test_nn.py]: Same here. You could put the graph you'd like to test here: https://github.com/dmlc/dgl/blob/master/tests/test_utils/graph_cases.py

File: python/dgl/nn/mxnet/conv/gatedgraphconv.py [hunk 3]
```diff
@@ -75,6 +75,8 @@ def forward(self, graph, feat, etypes):
             The output feature of shape :math:`(N, D_{out})` where :math:`D_{out}`
             is the output feature size.
         """
+        assert graph.is_homograph, \
+            "Graph is not homogeneous; store edge types as features instead."
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/nn/mxnet/conv/gatedgraphconv.py]: The edge types should be stored as edge features?

File: python/dgl/heterograph.py [hunk 4]
```diff
@@ -4077,6 +4077,11 @@ def local_scope(self):
         self._node_frames = old_nframes
         self._edge_frames = old_eframes
 
+    @property
+    def is_homograph(self):
+        """Return if the graph is homogeneous."""
+        return len(self.ntypes) == 1 and len(self.etypes) == 1
```
[the remark this answers is not in the document]
Comment (yzh119) [reply python/dgl/heterograph.py]: agree

File: python/dgl/nn/mxnet/conv/edgeconv.py [hunk 5]
```diff
@@ -60,17 +61,23 @@ def forward(self, g, h):
         h : mxnet.NDArray
             :math:`(N, D)` where :math:`N` is the number of nodes and
             :math:`D` is the number of feature dimensions.
+
+            If a pair of tensors is given, the graph must be a uni-bipartite graph
+            with only one edge type, and the two tensors must have the same
+            dimensionality on all except the first axis.
```
[the remark this answers is not in the document]
Comment (yzh119) [reply python/dgl/nn/mxnet/conv/edgeconv.py]: @BarclayII , I have checked the shape for gcn aggregator in SAGEConv, do we have other modules that requires equal shape?

File: python/dgl/nn/mxnet/conv/gatedgraphconv.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/nn/mxnet/conv/gatedgraphconv.py]: Is edge type required by this NN module? Maybe just `The XXX module only supports homogeneous graph.`?

[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/nn/mxnet/conv/gatedgraphconv.py]: Yeah you're right.  We don't need to actually store the edge type as a feature.  However this module works on a graph with edge types.  I guess I can change the error message to "This module only supports homogeneous graph; please specify edge types as an argument"?

File: python/dgl/nn/mxnet/conv/edgeconv.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/nn/mxnet/conv/edgeconv.py]: I don't think so.

**Pull request merged**